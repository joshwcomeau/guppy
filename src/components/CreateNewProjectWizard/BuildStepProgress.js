// @flow
import React, { PureComponent } from 'react';
import styled from 'styled-components';
import IconBase from 'react-icons-kit';
import { check } from 'react-icons-kit/feather/check';

import Spinner from '../Spinner';
import { COLORS } from '../../constants';
import FadeIn from '../FadeIn';

type Status = 'upcoming' | 'in-progress' | 'done';

type Props = {
  step: any,
  status: Status,
};

type State = {
  shouldShowAdditionalCopy: boolean,
};

class BuildStepProgress extends PureComponent<Props, State> {
  timeoutId: number;

  state = {
    shouldShowAdditionalCopy: false,
  };

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.status === 'done') {
      this.setState({ shouldShowAdditionalCopy: false });
    }
  }

  componentDidUpdate(prevProps: Props) {
    if (
      prevProps.status !== 'in-progress' &&
      this.props.status === 'in-progress'
    ) {
      this.timeoutId = window.setTimeout(this.triggerAdditionalCopy, 5000);
    }
  }

  componentWillUnmount() {
    window.clearTimeout(this.timeoutId);
  }

  triggerAdditionalCopy = () => {
    this.setState({ shouldShowAdditionalCopy: true });
  };

  render() {
    const { step, status } = this.props;
    const { shouldShowAdditionalCopy } = this.state;
    return (
      <Wrapper isUpcoming={status === 'upcoming'}>
        <IconWrapper>
          {status === 'in-progress' && <Spinner size={20} />}
          {status === 'done' && <Checkmark size={32} icon={check} />}
        </IconWrapper>
        <ChildWrapper>
          <MainCopy>{step.copy}</MainCopy>
          {step.additionalCopy &&
            shouldShowAdditionalCopy && (
              <FadeIn>
                <AdditionalCopy>{step.additionalCopy}</AdditionalCopy>
              </FadeIn>
            )}
        </ChildWrapper>
      </Wrapper>
    );
  }
}

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  padding: 10px;
  font-size: 18px;
  opacity: ${props => (props.isUpcoming ? 0.5 : 1)};
  will-change: opacity;
  transition: opacity 800ms;
`;

const IconWrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 32px;
  height: 32px;
  margin-right: 10px;
`;

const ChildWrapper = styled.div`
  position: relative;
  flex: 1;
`;

const Checkmark = styled(IconBase)`
  color: ${COLORS.lightSuccess};
`;

const MainCopy = styled.div`
  font-weight: 600;
  -webkit-font-smoothing: antialiased;
  font-size: 18px;
`;

const AdditionalCopy = styled.div`
  position: absolute;
  top: 110%;
  font-size: 14px;
`;

export default BuildStepProgress;
