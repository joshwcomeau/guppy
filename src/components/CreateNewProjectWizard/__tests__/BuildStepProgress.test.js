import React from 'react';
import { shallow } from 'enzyme';

import BuildStepProgress from '../BuildStepProgress';

describe('BuildStepProgress component', () => {
  let wrapper;
  let instance;

  const mockStep = {
    copy: 'Building...',
    additionalCopy: 'This may take some time',
  };

  jest.useFakeTimers();

  beforeEach(() => {
    wrapper = shallow(<BuildStepProgress step={mockStep} status="upcoming" />);
    instance = wrapper.instance();
  });

  it('should render upcoming', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render in-progress', () => {
    wrapper.setProps({ status: 'in-progress' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render done', () => {
    wrapper.setProps({ status: 'done' });
    expect(wrapper).toMatchSnapshot();
  });

  describe('Component logic', () => {
    it('should hide additional copy if done', () => {
      wrapper.setState({
        shouldShowAdditionalCopy: true,
      });
      wrapper.setProps({
        status: 'done',
      });
      expect(instance.state.shouldShowAdditionalCopy).toBeFalsy();
    });

    it('should show additonal copy after a delay', () => {
      wrapper.setProps({
        status: 'in-progress',
      });
      jest.runAllTimers();
      expect(instance.state.shouldShowAdditionalCopy).toBeTruthy();
    });

    it('should clear timeout on unmount', () => {
      window.clearTimeout = jest.fn();
      instance.componentWillUnmount();
      expect(window.clearTimeout).toHaveBeenCalledWith(instance.timeoutId);
    });
  });
});
