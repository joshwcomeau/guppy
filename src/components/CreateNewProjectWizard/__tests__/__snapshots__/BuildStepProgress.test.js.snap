// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildStepProgress component should render done 1`] = `
<styled.div
  isUpcoming={false}
>
  <styled.div>
    <Styled(Icon)
      icon={
        Object {
          "attribs": Object {
            "fill": "none",
            "stroke": "currentColor",
            "stroke-linecap": "round",
            "stroke-linejoin": "round",
            "stroke-width": "2",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "points": "20 6 9 17 4 12",
              },
              "children": Array [],
              "name": "polyline",
            },
          ],
          "viewBox": "0 0 24 24",
        }
      }
      size={32}
    />
  </styled.div>
  <styled.div>
    <styled.div>
      Building...
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`BuildStepProgress component should render in-progress 1`] = `
<styled.div
  isUpcoming={false}
>
  <styled.div>
    <Spinner
      size={20}
    />
  </styled.div>
  <styled.div>
    <styled.div>
      Building...
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`BuildStepProgress component should render upcoming 1`] = `
<styled.div
  isUpcoming={true}
>
  <styled.div />
  <styled.div>
    <styled.div>
      Building...
    </styled.div>
  </styled.div>
</styled.div>
`;
